weather_dl/download_pipeline/manifest.py

```diff
@@ -102,80 +199,140 @@ class Manifest(abc.ABC):
-            download_duration=None
-        )
+        self.status = dataclasses.replace(self._read(location))
+        self.status.config_name = config_name
+        self.status.selection = selection
+        self.status.location = location
+        self.status.user = user
 
     def __enter__(self) -> None:
-        """Record 'in-progress' status of a transaction."""
-        self.start = time.time()
-        self._update(self.status)
+        pass
 
     def __exit__(self, exc_type, exc_inst, exc_tb) -> None:
         """Record end status of a transaction as either 'success' or 'failure'."""
-        end = time.time()
         if exc_type is None:
-            status = 'success'
+            status = Status.SUCCESS
             error = None
         else:
-            status = 'failure'
+            status = Status.FAILURE
             # For explanation, see https://docs.python.org/3/library/traceback.html#traceback.format_exception
             error = '\n'.join(traceback.format_exception(exc_type, exc_inst, exc_tb))
 
-        self.status = DownloadStatus(
-            selection=self.status.selection,
-            location=self.status.location,
-            user=self.status.user,
-            status=status,
-            error=error,
-            download_scheduled_time=self.status.download_scheduled_time,
-            download_finished_time=int(end),
-            download_duration=int(end - self.start)
+        new_status = dataclasses.replace(self.status)
+        new_status.error = error
+        new_status.status = status
+        current_utc_time = (
+            datetime.datetime.utcnow()
+            .replace(tzinfo=datetime.timezone.utc)
+            .isoformat(timespec='seconds')
         )
+
+        # This is necessary for setting the precise start time of the previous stage
+        # and end time of the final stage, as well as handling the case of Status.FAILURE.
+        if new_status.stage == Stage.FETCH:
+            new_status.fetch_start_time = self.prev_stage_precise_start_time
+            new_status.fetch_end_time = current_utc_time
+        elif new_status.stage == Stage.RETRIEVE:
+            new_status.retrieve_start_time = self.prev_stage_precise_start_time
+            new_status.retrieve_end_time = current_utc_time
+        elif new_status.stage == Stage.DOWNLOAD:
+            new_status.download_start_time = self.prev_stage_precise_start_time
+            new_status.download_end_time = current_utc_time
+        else:
+            new_status.upload_start_time = self.prev_stage_precise_start_time
+            new_status.upload_end_time = current_utc_time
+
+        path = new_status.location
+        parsed_gcs_path = urlparse(path)
+        if parsed_gcs_path.scheme != 'gs' or parsed_gcs_path.netloc == '':
+            new_status.size = LocalSystemFileSizeStrategy().get_file_size(path)
+        else:
+            new_status.size = GCSBlobSizeStrategy().get_file_size(parsed_gcs_path)
```

Comment:
This is much cleaner. However, could we apply something like the strategy pattern on the manifest subclass itself? e.g. the GCS Manifest has a `clac_file_size()` method that gets overridden? I see no need to create a separate class hierarchy, this one should do fine.

Reply:
For example, our `_read` and `_update` calls use the strategy pattern in my book.

Reply:
I changed the implementation a bit w.r.t to getting file size. WRYT ?